File: frontend/lib/model/judgement/userankjudgement/model/UseRankJudgement.ts
import UseRankInstance from "../../../instance/userankinstance/model/UseRankInstance";
import { IJudgement, IJudgementConstructor } from "../../model/IJudgement";
import UseRankJudgementDto from "../dto/UseRankJudgementDto";
import UseRankJudgementId from "./UseRankJudgementId";

export default class UseRankJudgement implements IJudgement {

    readonly id: UseRankJudgementId;

    readonly instance: UseRankInstance;
    
    readonly label: string;
    readonly comment: string;

    constructor(id: UseRankJudgementId, instance: UseRankInstance, label: string, comment: string) {
        this.id = id;
        this.instance = instance;
        this.label = label;
        this.comment = comment;
    }

    public getId(): UseRankJudgementId {
        return this.id;
    }

    public getInstance(): UseRankInstance {
        return this.instance;
    }

    public getLabel(): string {
        return this.label;
    }

    public getComment(): string {
        return this.comment;
    }

    public static fromDto(dto: UseRankJudgementDto): UseRankJudgement {
        return new UseRankJudgement(
            UseRankJudgementId.fromDto(dto.id),
            UseRankInstance.fromDto(dto.instance),
            dto.label,
            dto.comment
        );
    }

}

export class UseRankJudgementConstructor implements IJudgementConstructor {
    fromDto(dto: UseRankJudgementDto): UseRankJudgement {
        return UseRankJudgement.fromDto(dto);
    }
}

File: backend/src/main/java/de/garrafao/phitag/domain/judgement/userankjudgement/UseRankJudgement.java
package de.garrafao.phitag.domain.judgement.userankjudgement;

import de.garrafao.phitag.domain.annotator.Annotator;
import de.garrafao.phitag.domain.instance.userankinstance.UseRankInstance;
import de.garrafao.phitag.domain.judgement.IJudgement;
import lombok.Getter;
import lombok.Setter;
import org.apache.commons.lang3.Validate;

import javax.persistence.*;

@Entity
@Table(name = "phitaguserankjudgement")
@Getter
public class UseRankJudgement implements IJudgement {

    @Setter
    @EmbeddedId
    private UseRankJudgementId id;

    @MapsId("annotatorid")
    @ManyToOne
    private Annotator annotator;

    @MapsId("instanceid")
    @ManyToOne
    private UseRankInstance instance;

    @Setter
    @Column(name = "rank", nullable = false)
    private String  label;

    @Setter
    @Column(name = "comment", nullable = false)
    private String comment;

    public UseRankJudgement() {
    }

    public UseRankJudgement(UseRankInstance useRankInstance, Annotator annotator, String label, String comment) {
        Validate.notNull(useRankInstance, "instanceData must not be null");
        Validate.notNull(annotator, "annotator must not be null");
        Validate.notNull(label, "label must not be null");
        Validate.notNull(comment, "comment must not be null");

        this.id = new UseRankJudgementId(useRankInstance.getId(), annotator.getId());

        this.instance = useRankInstance;
        this.annotator = annotator;

        this.label = label;
        this.comment = comment;
    }

    public UseRankInstance getUseRankInstance() {
        return instance;
    }

    @Override
    public String toString() {
        return "UseRankJudgement [id=" + id + ", label=" + label + ", comment=" + comment + "]";
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj) {
            return true;
        }
        if (obj == null) {
            return false;
        }
        if (getClass() != obj.getClass()) {
            return false;
        }
        UseRankJudgement other = (UseRankJudgement) obj;
        return this.id.equals(other.getId());
    }

    @Override
    public int hashCode() {
        return this.id.hashCode();
    }

}
